bin/helpers/openstack.sh

```diff
@@ -0,0 +1,357 @@
+  local upload_filepath="/tmp/${image_name}.${disc_format}"
+  local image_list
+  local image_id
+  local image_details
+  local server_checksum
+  local host_checksum
+  image_id=$(get_image_id "${upload_name}")
+  server_checksum=$(get_server_uploaded_image_checksum "${image_id}")
+
+  if [[ ${#server_checksum} != 32 ]]; then
+    echo "No valid server_checksum field on server, skipping checksum verification step"
+  else
+    host_checksum=$(get_host_uploaded_image_checksum "${image_name}" "${disc_format}" "${image_url}")
+
+    if [ "${host_checksum}" != "${server_checksum}" ]; then
+      echo >&2 "Server checksum: ${server_checksum}"
+      echo >&2 "Expected checksum: ${host_checksum}"
+      echo >&2 "The server image checksum does not match expectations"
+      echo >&2 "Something might have failed on file transfer."
+      echo >&2 "Please delete image ${image_name} from Openstack and try again."
+      exit 1
+    else
+      echo "Uploaded image checksum is OK"
+    fi
+  fi
+}
+
+function upload_image() {
+  local image_name=$1
+  local disc_format=$2
+  local image_url=$3
+  local upload_name="${image_name}"
+  local upload_filepath="/tmp/${image_name}.${disc_format}"
+  local min_disc=5
+  local image_size
+  local image_size_number
+  image_size=$(du -sh --block-size=G --apparent-size "${upload_filepath}" | awk '{print $1}')
+  image_size_number=${image_size::-1}
+
+  if [[ "${disc_format}" == "raw" ]]; then
+    min_disc=${image_size_number};
+  fi
+
+  # Upload image
+  echo "Starting upload of ${upload_name}"
+  echo -e "\n"
+  echo "image source: ${upload_filepath}"
+  echo "image name: ${upload_name}"
+  echo "image format: ${disc_format}"
+  echo "image size: ${image_size_number}B"
+  echo -e "\n"
+  glance image-create \
+    --file "${upload_filepath}" \
+    --disk-format "${disc_format}" \
+    --min-disk "${min_disc}" \
+    --container-format bare \
+    --name "${upload_name}" \
+    --progress
+
+  echo "Verifying the uploaded image checksum"
+  echo -e "\n"
+
+  verify_uploaded_image "${image_name}" "${disc_format}" "${image_url}"
+}
+
+function verify_downloaded_image() {
+  local download_filename=$1
+  local download_url=$2
+  local download_filepath="/tmp/${download_filename}"
+
+  echo "Downloading ${download_url}/${download_filename}.md5 file"
+  curl "${download_url}/${download_filename}.md5" \
+    -o "${download_filepath}.md5" \
+    --connect-timeout 30 \
+    --max-time 30
+
+  # Verify md5sum of downloaded file
+  echo "Checking md5 sum"
+  md5result=$(
+    cd /tmp || exit;
+    md5sum -c "${download_filename}.md5"
+  )
+  if [[ "${md5result}" != *": OK"* ]]; then
+    echo >&2 "Wrong checksum of downloaded image."
+    echo >&2 "Something might have failed on file transfer."
+    echo >&2 "Please try again."
+    exit 1
+  else
+    echo "Checksum of downloaded image OK"
+  fi
+}
+
+function post_image_upload_checks() {
+  local image_name=$1
+  local disc_format=$2
```

Comment:
To avoid confusion with image format on image downloaded change "disk_format" to "glance_disk_format"